File: 4-DFS/DFS.py
class Node:

    def __init__(self,value):
        self.value = value
        self.children = []

    def addChild(self,node):
        self.children.append(node)

    
    def DFS(self):
        results = []
        return self.DFSHelper(results)

    
    def DFSHelper(self,results):

        results.append(self)

        for child in self.children:
            child.DFSHelper(results)
        
        return results



n1 = Node(1)
n2 =  Node(2)
n3 =  Node(3)
n4 =  Node(4)
n5 =  Node(5)
n6 =  Node(6)
n7 =  Node(7)
n8 =  Node(8)

n1.addChild(n2)
n1.addChild(n3)

n2.addChild(n4)

n3.addChild(n5)
n3.addChild(n6)

n6.addChild(n7)
n8.addChild(n7)


for n in n1.DFS():
    print(n.value)

File: 4-DFS/DFS.java
import java.util.ArrayList;

class NodeDFS {

    int value;
    ArrayList<NodeDFS> children;

    public NodeDFS(int value) {
        this.value = value;
        this.children = new ArrayList<>();
    }

    public void addChild(NodeDFS n) {
        this.children.add(n);
    }

    public ArrayList<NodeDFS> DFS() {
        // time O(E+V)
        // space O(V)
        ArrayList<NodeDFS> result = new ArrayList<>();

        return DFSHelper(result);
    }

    public ArrayList<NodeDFS> DFSHelper(ArrayList<NodeDFS> path) {

        path.add(this);

        for (NodeDFS child : this.children) {
            child.DFSHelper(path);
        }

        return path;
    }

}

class testDFS {

    public static void main(String args[]) {
        
        NodeDFS n1 = new NodeDFS(1);
        NodeDFS n2 = new NodeDFS(2);
        NodeDFS n3 = new NodeDFS(3);
        NodeDFS n4 = new NodeDFS(4);
        NodeDFS n5 = new NodeDFS(5);
        NodeDFS n6 = new NodeDFS(6);
        NodeDFS n7 = new NodeDFS(7);
        NodeDFS n8 = new NodeDFS(8);

        n1.addChild(n2);
        n1.addChild(n3);

        n2.addChild(n4);

        n3.addChild(n5);
        n3.addChild(n6);

        n6.addChild(n7);
        n8.addChild(n7);

        for (NodeDFS node : n1.DFS()) {
            System.out.print(node.value + " ");
        }
        System.out.println();
    }
}
